Implement GH-47: Add disabled state for button component

mewtaylor opened the pull request:
This implements #47 by adding a disabled state to the button component, and then by applying it to top row buttons for group/ungroup, and front/back forward/backward.

mewtaylor committed:
Add disabled state for button

This implements #47 by adding a disabled state to the button component, and then by applying it to top row buttons for group/ungroup, and front/back forward/backward.

paulkaplan commented:
For some reason things seem to initialize in a strange way. Until i drag something around, everything is disabled, even if something is selected.

![odd-select-behavior](https://user-images.githubusercontent.com/654102/32023046-2ba56c56-b9a6-11e7-9b55-5df78b66bde7.gif)

(this was recorded from within the GUI, fwiw)

fsih commented on src/components/paint-editor/paint-editor.jsx:
```diff
@@ -62,126 +63,134 @@ class PaintEditorComponent extends React.Component {
+                                <ButtonGroup>
+                                    <Button
+                                        className={styles.buttonGroupButton}
+                                        onClick={this.props.onUndo}
```
Not necessarily in this PR, but we should add disabled states for undo and redo too. (The functions aren't pre-existing in helper/undo, but if you look in reducers/undo, they would look like those functions that log "Can't undo" and "can't redo")

fsih replied:
canUndo: return (state.pointer > 1)
canRedo: return (state.pointer > -1 && state.pointer !== state.stack.length - 1)

mewtaylor commented:
@paulkaplan ah that makes sense – since selecting is not causing a re-render, it's not getting a new check. Probably makes sense to have the call to `should...()` come from a reducer, and be dispatched on a select state perhaps? Thoughts @fsih ?

paulkaplan commented:
Another issue, I seem to get a bunch of these proptype warnings:
```
warning.js:33 Warning: Failed prop type: Invalid prop `disabled` of type `boolean` supplied to `ButtonComponent`, expected `string`.
```

fsih commented:
@mewtaylor That makes sense. Should show group/ungroup/front/back can listen on state.scratchPaint.selectedItems, which should update when the selection updates. I don't think it currently pushes a state change if you reorder items, so we will have to add that (calling setSelectedItems in bring to front/back etc.)

mewtaylor committed:
Add bool to `disabled` prop type

fixes a warning that @paulkaplan noticed, thanks!

mewtaylor committed:
Have the paint-editor container listen to `selectedItems`

This solves the issue of buttons not updating their disabled/enabled state until being dragged by making sure that a re-render occurs whenever a new item is selected, checking again for state.

mewtaylor commented:
@paulkaplan I just pushed some changes to address the warning you were seeing, as well as the behavior issues you noticed too – thanks for that!

@fsih I'm happy to add enabled/disabled for undo/redo in this PR too, makes sense to me!

mewtaylor committed:
Add undo/redo disabled states

mewtaylor committed:
pass props to button from labeled button

so that we don’t have to check for things like `disabled` twice.

mewtaylor commented:
@paulkaplan @fsih I added the undo/redo disabled checks, lemme know what you think!

fsih reviewed (dismissed):
Looks great!

Just a small nit, when redo is enabled and undo is disabled, should the line between them be grayed out?
![image](https://user-images.githubusercontent.com/2855464/32062952-fd663e44-ba43-11e7-84cd-e1a2e0e3d49b.png)

mewtaylor committed:
Change border on disabled undo/enabled redo

thanks @fsih

mewtaylor committed:
comment out disabled mode tools for now
